### ISSUE TITLE
is there a way to prevent routes from triggering?

### ISSUE THREAD

**Issue by awei01** [cerebral/cerebral-module-router#93]
Hi, I'm wondering if there's a way to prevent routes from triggering or addressbar url change. 

Use case: Someone is filling out a form and doesn't save. But then they click on a different link or change address in address bar. Is there a best-practice way to handle this?

**Comment by Guria** [comment]
@awei01 great catch. fix is on the way, thanks

**Comment by Guria** [comment]
I can't see a good way to prevent such case. Link component triggers signals, not routes. You can add an action to your signal check which will check that navigation currently unwanted. But from UX point of view I would recommend just have smart form state restore if user returns back after unwanted navigation.

**Comment by christianalfoni** [comment]
I would suggest the same approach as "authentication" again.

```js
module.addSignals({
  homeClicked: preventUnfinished(homeClicked)
})
```
This factory has the logic for ensuring that something started is also finished:

```js
function preventUnfinished(chain) {
  isNotFinished, {
    formA: [notifyFormAError],
    formB: [notifyFormBError],
    otherwise: chain
  }
}
```

This will actually also notify when the url is manually changed... which is not the case typically in these scenarios. Does that make sense?

**Comment by awei01** [comment]
This makes sense. But, now it seems like any routeable signal will require these factories to work properly. 

Just throwing out another idea: What if the `cerebral-module-router` had a services like `preventRouting` and `stopPreventRouting` which you could turn on and off. Then in my module which manages the form state, i can call `preventRouting` upon form change and then `stopPreventRouting` upon form completion?

**Comment by christianalfoni** [comment]
@awei01 That is an interesting suggestion! Though I believe you would still need to use factories on all the routed signals, because I suppose you want to give an error if the user tries to route and it is prevented? Hm hm :)

**Comment by awei01** [comment]
Good point... I'll play around with your suggestion. If I find anything useful, I'll post it here.

**Comment by awei01** [comment]
Took some of @christianalfoni's suggestions and created a POC here: https://github.com/awei01/cerebral-module-router-POC
